### PR TITLE
Issue #6447 - Deprecate support for UTF16 encoding in URIs

### PR DESCRIPTION
**Pull request by lachlan-roberts**
## Issue #6447

Deprecate support for UTF16 encoding in URIs and add compliance mode to allow UTF16 encodings.

### CODE REVIEW

File: jetty-http/src/main/java/org/eclipse/jetty/http/HttpURI.java [hunk 1]
```diff
@@ -73,9 +73,16 @@
...
-        PARAM
+        PARAM,
+
+        /**
+         * Contains UTF16 encodings
+         */
+        UTF16
     }
 
+    EnumSet<Violation> AMBIGUOUS = EnumSet.complementOf(EnumSet.of(Violation.UTF16));
```
Comment (joakime) [inline jetty-http/src/main/java/org/eclipse/jetty/http/HttpURI.java]: Is the name `AMBIGUOUS` here still valid?

Comment (joakime) [reply jetty-http/src/main/java/org/eclipse/jetty/http/HttpURI.java]: In fact, is this EnumSet even used? (I don't see it)

File: jetty-util/src/main/java/org/eclipse/jetty/util/URIUtil.java [hunk 2]
```diff
@@ -470,8 +470,7 @@ public static String decodePath(String path, int offset, int length)
                             char u = path.charAt(i + 1);
                             if (u == 'u')
                             {
-                                // TODO remove %u support in jetty-10
-                                // this is wrong. This is a codepoint not a char
+                                // UTF16 encoding is only supported with UriCompliance.Violation.UTF16_ENCODINGS
```
Comment (joakime) [inline jetty-util/src/main/java/org/eclipse/jetty/util/URIUtil.java]: What about the comment about removing %u support on line 560 in this file too?

https://github.com/eclipse/jetty.project/blob/a09d9e94b5cc542b8436f9ef3171da42cfc3b13b/jetty-util/src/main/java/org/eclipse/jetty/util/URIUtil.java#L560-L562

**Review by joakime — CHANGES_REQUESTED** [review]
Minor nits about unused fields and javadoc comments.

File: jetty-util/src/main/java/org/eclipse/jetty/util/URIUtil.java (hunk 2 above)
Comment (joakime) [reply jetty-util/src/main/java/org/eclipse/jetty/util/URIUtil.java]: This comment doesn't seem to apply here, as `UriCompliance.Violation.UTF16_ENCODINGS` are not even used in this URIUtil class?

File: jetty-http/src/main/java/org/eclipse/jetty/http/HttpURI.java (hunk 1 above)
Comment (gregw) [reply jetty-http/src/main/java/org/eclipse/jetty/http/HttpURI.java]: It is still the right name, as it is the subset that is ambiguous.  It was used in a version... but not now... Unless there is an efficient way to get the intersection of 2 enum sets

File: jetty-http/src/main/java/org/eclipse/jetty/http/HttpURI.java [hunk 3]
```diff
@@ -1033,8 +1080,9 @@ private void parse(State state, final String uri)
                                 break;
                             case '%':
                                 // must have be in an encoded path
```
Comment (janbartel) [inline jetty-http/src/main/java/org/eclipse/jetty/http/HttpURI.java]:
```suggestion
                                // must have been in an encoded path
```

File: jetty-http/src/main/java/org/eclipse/jetty/http/HttpURI.java [hunk 4]
```diff
@@ -938,9 +982,11 @@ private void parse(State state, final String uri)
             int mark = 0; // the start of the current section being parsed
             int pathMark = 0; // the start of the path section
             int segment = 0; // the start of the current segment within the path
-            boolean encoded = false; // set to true if the path contains % encoded characters
+            boolean encodedPath = false; // set to true if the path contains % encoded characters
+            boolean encodedUtf16 = false; // Is the current encoding for UTF16?
+            int encodedCharacters = 0; // partial state of parsing a % encoded character<x>
+            int encodedValue = 0; // the partial encoded value
             boolean dot = false; // set to true if the path containers . or .. segments
```
Comment (janbartel) [inline jetty-http/src/main/java/org/eclipse/jetty/http/HttpURI.java]:
```suggestion
            boolean dot = false; // set to true if the path contains . or .. segments
```

File: jetty-http/src/test/java/org/eclipse/jetty/http/HttpURITest.java [hunk 5]
```diff
@@ -323,109 +327,134 @@ public void testCanonicalDecoded()
         return Arrays.stream(new Object[][]
             {
                 // Simple path example
-                {"http://host/path/info", "/path/info", EnumSet.noneOf(Ambiguous.class)},
-                {"//host/path/info", "/path/info", EnumSet.noneOf(Ambiguous.class)},
-                {"/path/info", "/path/info", EnumSet.noneOf(Ambiguous.class)},
+                {"http://host/path/info", "/path/info", EnumSet.noneOf(Violation.class)},
```
Comment (janbartel) [inline jetty-http/src/test/java/org/eclipse/jetty/http/HttpURITest.java:330]: Unrelated to UTF16, but we don't seem to have any tests where the scheme name contains a ".",  which is a legal character.
Or percent encoded host names?

File: jetty-http/src/main/java/org/eclipse/jetty/http/HttpURI.java (hunk 1 above)
Comment (lachlan-roberts) [reply jetty-http/src/main/java/org/eclipse/jetty/http/HttpURI.java]: To work out if we have an ambiguous violation we do
```java
!_violations.isEmpty() && !(_violations.size() == 1 && _violations.contains(Violation.UTF16))
```

So we never use this `AMBIGUOUS` except in testing. Should I inline its usage and remove it?

File: jetty-http/src/test/java/org/eclipse/jetty/http/HttpURITest.java (hunk 5 above)
Comment (lachlan-roberts) [reply jetty-http/src/test/java/org/eclipse/jetty/http/HttpURITest.java:330]: I have added some extra tests for this.

**Review by joakime — APPROVED** [review]

**Pull request merged**